### DISCUSSION TITLE
创建Collection或者partition的时候使用太多内存

### DISCUSSION THREAD

**Discussion by susoooon** [milvus-io/milvus#16680] [Q&A and General discussion]
我发现我在使用2.0.2的时候，创建collection或者partition的时候会消耗很多内存，比如一共42个collection ,总数据不到10W，消耗了9G内存。
有什么办法解决这个问题吗？或者可以怎么设置预加载内存大小之类的

**Comment by yhmo** [comment]
我在单个物理机上以cluster模式启动了一个milvus服务，客户端创建50个collection，每个collection插入2000条128维向量，总数是10万条。观察各个milvus进程的内存使用，query node节点用的最多使用了2GB，其他进程使用量从几十MB到几百MB不等，总的使用量是3.2GB左右。另外，milvus所依托的第三方服务比如pulsar/minio/etcd，我用docker stats来察看，pulsar使用最多达到1.8GB，minio和etcd都在几十MB左右。所以总的使用量大约是5GB。
plusar是第三方服务，我们没法控制plusar的内存使用，但是它确实蛮吃内存。query node的内存使用确实有点超出预期，这里面原因有点复杂，milvus底层依赖C++的库，上层是用golang，数据结构在两种语言间传递时有非预期的内存消耗，相关开发人员在调研中。

**Comment by susoooon** [comment]
[milvus_memory_issue.txt](https://github.com/milvus-io/milvus/files/8580018/milvus_memory_issue.txt)
这是我的测试代码，我发现在跑这个代码的时候，无论部署在GPU还是CPU的不同服务器上都会导致内存随着插入逐渐变大，到占用过大以至于超过32G，OOM，我们得到的结果不一样，不知道是否和你的测试代码有区别。可以帮忙看一下问题出在哪里吗？万分感谢。

**Reply by yhmo** [reply]
这个脚本只建了1个collection，而partition有600多个，跑完这个脚本之后，query node的内存占用量确实比较离谱，我这里显示占用120GB。
我建了一个issue：https://github.com/milvus-io/milvus/issues/16711    相关开发人员会去看。

**Comment by yhmo** [comment]
@susoooon 
你的配置文件 milvus.yml 有没有什么改动？是用的默认配置么？

**Reply by susoooon** [reply]
是的，默认配置

**Comment by yhmo** [comment]
query node在加载数据时，会预先给每个segemnt分配一个空间，默认是32768行，在milvus.yml里定义：
```
queryNode:
  segcore:
    chunkRows: 32768 # The number of vectors in a chunk.
```

由于这个collection里有600个partition，默认shard number=2，于是每个partition里会有2个segemnt，于是querynode会构建1200个segment，每个segemnt里预留了32768条数据的空间，每条数据主要是768维的向量。所以就会占用这么多内存：600 * 2 * 32768 * 768 * 4 = 120795955200 bytes = 112GB

确实不太合理，目前你可以通过改变milvus.yml里那个chunkRows的大小来降低预分配的空间大小。比如改成64试试。我们后续会考虑做些优化。

**Reply by susoooon** [reply]
非常感谢，我现在暂时的方案是弃用partition，把它作为一个混合搜索的参数搜索。chunkRows 也可以更改collection的预分配空间吗？因为多个collection也有同样的问题。
再次，非常感谢！

**Reply by yhmo** [reply]
chunkRows是针对segment的，多个collection也一样

**Reply by susoooon** [reply]
> chunkRows是针对segment的，多个collection也一样

我试了一下，按照64的方案，存在不同collection仍然会内存爆炸，存在不同partition是没问题的。collection可能有其他配置可以像修改chunkRows一样暂时处理这个问题，不知道你是否有好的建议

**Reply by yhmo** [reply]
你分成了多少个collection？

**Reply by susoooon** [reply]
> 你分成了多少个collection？

修改chunkRows为32之后，只是把上面存Partition改为存collection，大约200多个Collection内存就十多G了

**Comment by zlw18813081566** [comment]
你好，内存增长这个问题你解决了吗？我也碰到了一样的情况

**Reply by susoooon** [reply]
2.1版本最近发布，不确定是否还有这个问题，如果你有时间可以试一下，如果可以的话告诉我一下。
我改了代码逻辑，把本来需要分collection的分到partition里，分partiton里的用expr字段查询区分。同时改了chunkRows: 32768改到32或者16。运行正常

**Reply by zlw18813081566** [reply]
好的，谢谢回复
